integrations/apollo/src/index.ts

```diff
@@ -56,19 +56,38 @@ export const createApolloClient = ({
   const uri = backendUrl
   const interpreter = nhost?.auth.client.interpreter
 
-  let token: string | null = null
+  let accessToken: AuthContext['accessToken'] | null = null
 
-  function getAuthHeaders() {
+  const isTokenValid = () =>
+    accessToken?.value && accessToken?.expiresAt && accessToken.expiresAt > new Date()
+
+  const awaitValidTokenOrNull = () =>
+    new Promise((resolve) => {
+      // doing this as an interval to avoid race conditions that I imagine can happen if listening to token changes. Maybe there's a better way.
+      const interval = setInterval(() => {
```

Comment:
Instead of this interval-based approach, maybe we could use something like this:

```ts
new Promise((resolve) => {
  interpreter?.onTransition(async (_state, event) => {
    if (event.type === 'TOKEN_CHANGED') {
      resolve(true)
    }
  })
})
```

if I understand this piece of code correctly, all we care about here is the time when the local `accessToken` variable changes

Reply:
Yes, but will TOKEN_CHANGED trigger when we get signed out as well or should we listen to SIGNED OUT as well? Cause i think we want the request to continue then as well but with the public role.

Wouldn't we want to clear this listener after we're done btw?

Reply:
Yes, the `TOKEN_CHANGED` event should be sent when the user is signed out.

We should clear the listener for sure. I just wanted to sketch out an alternative approach.

Reply:
Looking at this a bit more. Wouldn't this risk the onTransition in awaitValidTokenOrNull to be fired before the token variable is set in the other onTransition in this file? Then you might still use the invalid token even after the TOKEN_CHANGED have fired inside awaitValidTokenOrNull.

If this approach is implemented, then maybe we should just return the token directly from the onTransition state in awaitValidTokenOrNull and use that? Then you could also remove the need to store token in that let variable. But since it is a bit bigger changes I'm gonna leave that up to you.

I'm quite happy with the interval checking thing, since I don't want to make too big changes to this file.

Reply:
I just checked the state transitions. It is probably not the safest way to determine the latest access token so we're better off with this interval-based approach for now.